handlers/handlers.go

```diff
@@ -1,238 +1,238 @@
+			loginForm.EmailAddress, 
+			loginForm.Password, 
+		); err != nil { 
+			panic(err) 
+		} 
+ 
+		// Set our cookie to have a valid JWT Token as the value 
+		{ 
+			userId, err := userservice.GetIdForUserWithEmailAddress(loginForm.EmailAddress) 
+			if err != nil { 
+				panic(err) 
+			} 
+ 
+			token, err := createTokenAsString(userId, credentialTimeoutInMinutes) 
+			if err != nil { 
+				panic(err) 
+			} 
+ 
+			expirationTime := time.Now().Add(credentialTimeoutInMinutes * time.Minute) 
+ 
+			cookie := http.Cookie{ 
+				Name:     cerealNotesCookieName, 
+				Value:    token, 
+				Expires:  expirationTime, 
+				HttpOnly: true, 
+			} 
+ 
+			http.SetCookie(responseWriter, &cookie) 
+		} 
+ 
+		responseWriter.WriteHeader(http.StatusCreated) 
+		responseWriter.Write([]byte(fmt.Sprint("passward email combo was correct"))) 
+ 
+	case http.MethodDelete: 
+		// Cookie will overwrite existing cookie then delete itself 
+		cookie := http.Cookie{ 
+			Name:     cerealNotesCookieName, 
+			Value:    "", 
+			HttpOnly: true, 
+			MaxAge:   -1, 
+		} 
+ 
+		http.SetCookie(responseWriter, &cookie) 
+		responseWriter.WriteHeader(http.StatusOK) 
+		responseWriter.Write([]byte(fmt.Sprint("user succefully logged out"))) 
+	default: 
+		respondWithMethodNotAllowed( 
+			responseWriter, 
+			[]string{http.MethodPost, http.MethodDelete}) 
+	} 
+} 
+ 
+type AuthentictedRequestHandlerType func( 
+	http.ResponseWriter, 
+	*http.Request, 
+	models.UserId) 
+ 
+func AuthenticateOrRedirectToLogin( 
+	authenticatedHandlerFunc AuthentictedRequestHandlerType, 
+) func(http.ResponseWriter, *http.Request) { 
+	return func(responseWriter http.ResponseWriter, request *http.Request) { 
+			if userId, err := getUserIdFromJwtToken(request); err != nil { 
+				// If not loggedin redirect to login page 
+				http.Redirect( 
+					responseWriter, 
+					request, 
+					paths.LoginOrSignupPath, 
+					http.StatusTemporaryRedirect) 
+			} else { 
+				authenticatedHandlerFunc(responseWriter, request, userId) 
+			} 
+		} 
+} 
+ 
+// Authenticated Handlers 
+func HandleHomeRequest( 
+	responseWriter http.ResponseWriter, 
+	request *http.Request, 
+	userId models.UserId, 
+) { 
+	switch request.Method { 
+	case http.MethodGet: 
+		parsedTemplate, err := template.ParseFiles("templates/home.tmpl") 
+		if err != nil { 
+			log.Fatal(err) 
+		} 
+ 
+		parsedTemplate.Execute(responseWriter, userId) 
+	default: 
+		respondWithMethodNotAllowed(responseWriter, []string{http.MethodGet}) 
+	} 
+} 
+ 
+// UTIL 
+func respondWithMethodNotAllowed( 
+	responseWriter http.ResponseWriter, 
+	allowedMethods []string, 
+) { 
+	allowedMethodsString := strings.Join(allowedMethods, ", ") 
+	responseWriter.Header().Set("Allow", allowedMethodsString) 
+ 
+	statusCode := http.StatusMethodNotAllowed 
+	http.Error(responseWriter, http.StatusText(statusCode), statusCode) 
+} 
+ 
+func getRequestBody(request *http.Request) []byte { 
+	body, err := ioutil.ReadAll(request.Body) 
```

Comment:
This line is reading all the bytes from the `io.Reader` into memory. When possible, avoid reading all the bytes from an `io.Reader` into memory, instead give the `io.Reader` directly to whatever wants to consume it. That way the consumer can decide if it needs to read the whole byte stream, and even if it does need to read the whole byte stream it can decide if it needs all the bytes in memory at the same time or if it can stream it.

It looks like the bytes from `getRequestBody` are only being consumed by `json.Unmarshal`. I'd get rid of `getRequestBody()` and do the  so instead you could do

```go
func HandleUserRequest(...) {
   defer func() {
       io.CopyN(ioutil.Discard, request.Body, 512) 
       request.Body.Close()
   }()
   
   ...

   signupForm := &SignupForm{}  // alternatively signupForm := new(SignupForm)
   if err := json.NewDecoder(request.Body).Decode(signupForm); err != nil {
       ...
   }
}
```

Reply:
As a side note, i'm not 100% sure why you want to drain the reader before closing it. Someone told me to do it once. I've also seen this in go-github, which is a github client written in go that we use at work. I really like their code, and I copy a lot of their patterns. 

I chose the number 512 because that's what go-github does. I'm not sure why they do that and tbh 512 seems pretty low. I think it probably makes more sense to do `io.Copy(ioutil.Discard, request.Body)`, idk why go-github doesn't do that.

I _think_ the reason you want to drain the request.Body is because the drain will wait for the guy on the other end of the tcp connection to finish sending all of his data before terminating the tcp connection. If you Close before draining you risk closing the connection before the other guy is done.

Actually, now that I write this out, that's totally why you want to drain before closing.

It's worth noting that in this example, the drain is technically unnecessary since you're always calling `json.NewDecoder.Decode` on `request.Body` which will drain it for you ... but that's an implementation detail of the json package that you shouldn't depend on.

Reply:
@ethomas2 could you explain your code snippet above? I don't understand what's happening in the deferred function

Reply:
It takes up to 512 bytes from the request body and sends it to the equivilent of /dev/null. 

I used 512 instead of just discarding all the bytes because that's what go-github does. It probably makes more sense to discard all the bytes with `io.Copy`

All of that is irrelevant though. Apparently you don't have to drain the request body after go1.7